Feat/animations for number 220

drakkomaximo opened the pull request:
# Number Selection Animations Feature

Close #220 

## 📌 Description 
This pull request implements enhanced animations for number selection in the lottery ticket buying interface. The feature improves user experience by providing smooth, engaging animations when selecting, deselecting, and generating random numbers.

## 🎯 Motivation and Context 
The previous implementation had abrupt transitions when selecting numbers, which created a poor user experience. Users needed:
- Smooth visual feedback when selecting/deselecting numbers
- Engaging animations to draw attention to selected numbers
- Clear indication when reaching the 5-number limit
- Visual feedback for random number generation

This feature addresses the need for better user interaction feedback and creates a more engaging lottery ticket selection experience.

## 🛠️ How to Test the Change
1. 🔹 Navigate to `/buy-tickets` page
2. 🔹 Click on different numbers to see selection animations
3. 🔹 Try selecting more than 5 numbers to see limit reached animation
4. 🔹 Use the "Random" button for individual tickets to see random generation animations
5. 🔹 Use "Random for All" to see staggered animations across multiple tickets
6. 🔹 Verify that unselected options remain visible after selection

## 🖼️ Screenshots (if applicable) 


https://github.com/user-attachments/assets/a8dab796-e984-456b-ba9c-db3562b5a53f


https://github.com/user-attachments/assets/d5cd1fb7-af9c-46e5-bcf4-180d66bb6402

The feature includes:
- Bounce animation when selecting numbers (scale: 1 → 1.3 → 1.1)
- Shrink animation when deselecting numbers (scale: 1.1 → 0.8 → 1)
- Red flash animation when trying to select beyond 5 numbers
- Enhanced random generation with staggered animations
- Smooth color transitions for all states

## 🔍 Type of Change
- [x] ✨ **New Feature** - Adds new animation functionality
- [x] 🎨 **UX/UI Enhancement** - Improves user experience with animations

## ✅ Checklist Before Merging
- [x] 🧪 I have tested the code and it works as expected.
- [x] 🎨 My changes follow the project's coding style.
- [x] 📖 I have updated the documentation if necessary.
- [x] ⚠️ No new warnings or errors were introduced.
- [x] 🔍 I have reviewed and approved my own code before submitting.

## 📌 Additional Notes 

### Files Modified:
- `packages/nextjs/app/buy-tickets/page.tsx` - Enhanced with animation logic

### Key Features Implemented:

1. **Selection Animations**:
   - Bounce effect when selecting numbers
   - Smooth scale transitions (1 → 1.3 → 1.1)
   - Color transitions from gray to purple

2. **Deselection Animations**:
   - Shrink effect when deselecting numbers
   - Smooth scale transitions (1.1 → 0.8 → 1)
   - Color transitions from purple to gray

3. **Limit Reached Animation**:
   - Red flash animation when trying to select beyond 5 numbers
   - Clear visual feedback for limit enforcement

4. **Random Generation Animations**:
   - Staggered animations for individual ticket random generation
   - Sequential animations for "Random for All" functionality
   - Enhanced bounce effect for randomly generated numbers

5. **State Management**:
   - Added `animatingNumbers` state to track animation states
   - Proper cleanup of animation states after completion
   - Type-safe animation state management

### Technical Implementation:

1. **Animation Variants**: Created comprehensive Framer Motion variants for all states
2. **State Management**: Added animation state tracking with proper TypeScript types
3. **Timing Control**: Implemented staggered animations with setTimeout
4. **Performance**: Optimized animations to prevent memory leaks

### Animation States:
- `unselected`: Default state for available numbers
- `selected`: Active state for chosen numbers
- `deselecting`: Animation when removing a number
- `limitReached`: Animation when trying to exceed limit
- `random`: Animation for randomly generated numbers

### Future Improvements:
1. **Sound Effects**: Add audio feedback for selections
2. **Haptic Feedback**: Implement vibration for mobile devices
3. **Accessibility**: Add reduced motion support
4. **Performance**: Implement animation pooling for better performance
5. **Customization**: Allow users to adjust animation speed

This implementation provides a solid foundation for engaging number selection animations while maintaining performance and accessibility standards.

davidmelendez commented:
Please remove the number 00 from the selectable numbers since lottery standards are from 1 to 40. At the contract level, this rule will also be modified with the issue #329

davidmelendez commented:
@drakkomaximo Please remove the number 00 from the selectable numbers since lottery standards are from 1 to 40. At the contract level, this rule will also be modified with the issue #329

drakkomaximo commented:
# Buy Tickets Page Refactoring Report

## Overview
This document outlines the comprehensive refactoring and improvements made to the buy tickets functionality, focusing on user experience, visual consistency, and code organization.

## Changes Made

### 1. Component Structure Refactoring

#### TicketSelector Component
- **Location**: `packages/nextjs/components/buy-tickets/TicketSelector.tsx`
- **Purpose**: Main container for individual ticket selection
- **Key Features**:
  - Modular design with separate components for different functionalities
  - Proper prop passing for animations and interactions
  - Clean separation of concerns

#### NumberGrid Component
- **Location**: `packages/nextjs/components/buy-tickets/NumberGrid.tsx`
- **Purpose**: Handles the number selection grid (01-40 + LUCK)
- **Improvements**:
  - Enhanced contrast and visibility
  - Consistent color scheme with site palette
  - Better state management for selections

#### LotteryDisplay Component
- **Location**: `packages/nextjs/components/buy-tickets/LotteryDisplay.tsx`
- **Purpose**: Displays selected numbers with interactive elements
- **Key Features**:
  - Conditional interaction based on selection state
  - Proper cursor management
  - Animation support for number reveals

### 2. Visual Design Improvements

#### Color Scheme Consistency
- **Before**: Mixed color palette with inconsistent styling
- **After**: Unified purple-based color scheme matching site theme
- **Changes**:
  - Selected numbers: `bg-gradient-to-br from-purple-600 via-purple-700 to-purple-800`
  - Unselected numbers: `bg-gray-700 text-gray-200`
  - Disabled numbers: `bg-gray-700 text-gray-400 opacity-60`

#### Contrast Enhancements
- **Number Grid**:
  - Improved background contrast for better readability
  - Added subtle borders for number definition
  - Enhanced text contrast for accessibility
- **Selected Numbers**:
  - Added purple border for better visual distinction
  - Gradient backgrounds for depth
  - Clear visual hierarchy

#### LUCK Button Styling
- **Maintained**: Yellow-orange gradient for special element
- **Features**: Animated pulse effect and proper positioning
- **Accessibility**: Clear visual distinction from regular numbers

### 3. Interaction Logic Improvements

#### Conditional Interactions
- **Selected Numbers Display**:
  - Question marks: No hover effects, no cursor pointer
  - Actual numbers: Full interactivity with hover effects
  - Proper state management for click handlers

#### Grid Number Interactions
- **Selection States**:
  - Available: Full interactivity with hover effects
  - Selected: Visual feedback with enhanced styling
  - Disabled: Clear visual indication of non-interactive state

#### Animation System
- **Framework**: Framer Motion integration
- **Features**:
  - Smooth transitions between states
  - Staggered animations for multiple selections
  - Proper cleanup of animation states

### 4. Code Organization

#### Modular Component Architecture
```
buy-tickets/
├── TicketSelector.tsx      # Main ticket container
├── NumberGrid.tsx          # Number selection grid
├── LotteryDisplay.tsx      # Selected numbers display
├── TicketControls.tsx      # Ticket count controls
└── PurchaseSummary.tsx     # Purchase confirmation
```

#### State Management
- **Centralized State**: Main page manages all ticket states
- **Prop Drilling**: Clean prop passing to child components
- **Event Handling**: Consistent event handling patterns

#### Animation Variants
- **Number Selection**: Bounce and scale effects
- **Lottery Reveal**: Rotating and scaling animations
- **State Transitions**: Smooth color and opacity changes

### 5. User Experience Enhancements

#### Visual Feedback
- **Hover States**: Clear indication of interactive elements
- **Selection States**: Distinct visual styling for selected numbers
- **Disabled States**: Proper visual indication of non-interactive elements

#### Accessibility Improvements
- **Cursor Management**: Proper cursor states for different interactions
- **Color Contrast**: Enhanced contrast ratios for better readability
- **State Indication**: Clear visual feedback for all interaction states

#### Responsive Design
- **Grid Layout**: Responsive 7-column grid
- **Component Scaling**: Proper scaling for different screen sizes
- **Touch Interactions**: Optimized for mobile interactions

### 6. Technical Improvements

#### Performance Optimizations
- **Animation Cleanup**: Proper cleanup of animation states
- **Event Handling**: Efficient event delegation
- **State Updates**: Optimized state update patterns

#### Code Quality
- **TypeScript**: Proper type definitions for all components
- **Props Interface**: Clear interface definitions
- **Error Handling**: Proper error boundaries and fallbacks

#### Maintainability
- **Component Separation**: Clear separation of concerns
- **Reusable Patterns**: Consistent patterns across components
- **Documentation**: Clear code comments and structure

## Files Modified

### Core Components
1. `packages/nextjs/app/buy-tickets/page.tsx`
   - Main page logic and state management
   - Animation variant definitions
   - Component orchestration

2. `packages/nextjs/components/buy-tickets/NumberGrid.tsx`
   - Number selection grid
   - Visual styling improvements
   - Interaction logic

3. `packages/nextjs/components/buy-tickets/LotteryDisplay.tsx`
   - Selected numbers display
   - Conditional interaction logic
   - Animation integration

4. `packages/nextjs/components/buy-tickets/TicketSelector.tsx`
   - Individual ticket container
   - Component composition
   - Prop management

### Supporting Components
5. `packages/nextjs/components/buy-tickets/TicketControls.tsx`
   - Ticket count controls
   - Random generation functionality

6. `packages/nextjs/components/buy-tickets/PurchaseSummary.tsx`
   - Purchase confirmation interface
   - Transaction handling

## Testing Considerations

### Visual Testing
- [ ] Color contrast validation
- [ ] Responsive design testing
- [ ] Animation performance testing
- [ ] Cross-browser compatibility

### Functional Testing
- [ ] Number selection logic
- [ ] State management
- [ ] Animation cleanup
- [ ] Error handling

### Accessibility Testing
- [ ] Screen reader compatibility
- [ ] Keyboard navigation
- [ ] Color blind accessibility
- [ ] Focus management

## Future Enhancements

### Potential Improvements
1. **Advanced Animations**: More sophisticated animation sequences
2. **Accessibility**: Enhanced screen reader support
3. **Performance**: Further optimization of animation performance
4. **Testing**: Comprehensive unit and integration tests

### Technical Debt
1. **Type Definitions**: More specific TypeScript interfaces
2. **Error Boundaries**: Enhanced error handling
3. **Performance Monitoring**: Animation performance metrics
4. **Documentation**: Enhanced component documentation

## Conclusion

The refactoring successfully improved the buy tickets functionality with:
- **Better Visual Consistency**: Unified color scheme and styling
- **Enhanced User Experience**: Clear interaction states and feedback
- **Improved Code Organization**: Modular component architecture
- **Better Accessibility**: Proper contrast and interaction states
- **Maintainable Code**: Clean separation of concerns and reusable patterns

The implementation maintains the existing functionality while significantly improving the visual design, user experience, and code maintainability. 


https://github.com/user-attachments/assets/c60922e9-49b0-4abc-8b56-6a3cd38eb69d
